tests/table/test_upsert.py

```diff
@@ -511,6 +511,96 @@ def test_upsert_without_identifier_fields(catalog: Catalog) -> None:
         tbl.upsert(df)
 
 
+def test_upsert_struct_field_fails_in_join(catalog: Catalog) -> None:
```

Comment:
nit change the name since it no longer fails :) 

i also added another test [here](https://github.com/apache/iceberg-python/pull/1949/files#diff-0a8709ee278f0979ecdcff40f5ff7e215076867112c820b633e3e6f17c827a2eR557-R688) for when complex type is used as the join key

Reply:
I copied your test to this branch and renamed the non failing :-)